lib/matplotlib/axes/_base.py

```diff
@@ -2549,7 +2549,9 @@ def _process_unit_info(self, datasets=None, kwargs=None, *, convert=True):
             except KeyError:
                 raise ValueError(f"Invalid axis name: {axis_name!r}") from None
             # Update from data if axis is already set but no unit is set yet.
-            if axis is not None and data is not None and not axis.have_units():
+            if (axis is not None and
+                    data is not None and not
+                    axis._have_units_and_converter()):
```

Comment:
Is this only fixing symptoms but not the root cause?

At this position `plot()` results in `units` and `converter` being set, but `scatter()` has only set `units` and `converter` is still None. As little as I know about units - this discrepancy feels wrong.

Reply:
(it's been a while but...) I think that inside this `if` clause is designed to ensure that both units and a converter have been set, so it makes sense alter the logic to do this if either one is not set.